### PR TITLE
Update commands

### PR DESCRIPTION
**Pull request by onerandomusername**
Update several commands to use replies as arguments.
Update xkcd to be able to show the latest comic.

### CODE REVIEW

File: cogs/fun.py [hunk 1]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
 	@commands.command(hidden=True)
 	@commands.cooldown(rate=3, per=10.0, type=commands.BucketType.user)
 	@commands.bot_has_permissions(embed_links=True)
-	async def xkcd(self, ctx, *, id: int = None):
+	async def xkcd(self, ctx, id: typing.Union[int, str] = None):
```
Comment (SimplyMika) [inline cogs/fun.py:428]:
```py
Traceback (most recent call last):
  ...
  File "/home/ace/test/AceBot/cogs/fun.py", line 428, in Fun
    async def xkcd(self, ctx, id: typing.Union[int, str] = None):
NameError: name 'typing' is not defined
```
You have to include `typing`, [it's a standard module](https://docs.python.org/3/library/typing.html).
Also, I suggest doing `from typing import Union` just so that it's cleaner (you can then do `id: Union[int, str]`.

File: cogs/fun.py [hunk 2]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
+			if isinstance(id,int) and not id == 0:
+				# specific comic
+				id = abs(id)
+				url = f"https://xkcd.com/{id}/info.0.json"
+			elif id == 0 or isinstance(id,str) and id.lower() in ("latest","recent","today"): 
+				#latest comic
+				url = 'https://xkcd.com/info.0.json'
+			else: # id must be None or a str that isn't asking for today
+				#random comic
+				resp = await http.get('https://c.xkcd.com/random/comic')
```
Comment (SimplyMika) [inline cogs/fun.py]:
```py
Traceback (most recent call last):
  ...
  File "/home/ace/test/AceBot/cogs/fun.py", line 443, in xkcd
    resp = await http.get('https://c.xkcd.com/random/comic')
NameError: name 'http' is not defined
```
You probably meant `ctx.http.get` as in the original code.

You should also stick with the `async with ctx.http.get(url) as resp:` pattern, rather than assigning just using an assignment.

File: cogs/fun.py [hunk 3]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
-
 		async with ctx.typing():
-			async with ctx.http.get(url) as resp:
+			resp = None
+			comic_json = None
+			url = None
+			if isinstance(id,int) and not id == 0:
+				# specific comic
+				id = abs(id)
+				url = f"https://xkcd.com/{id}/info.0.json"
```
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
			if isinstance(id, int) and id > 0:
				# specific comic
				url = 'https://xkcd.com/{}/info.0.json'.format(id)
```
Negatives here don't make sense imo.
Also keep consistency:
* single quotes
* spaces after commas
* `''.format` rather than `f''` for compatibility (f-strings were introduced in Python 3.6)

File: cogs/fun.py [hunk 4]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
+			elif id == 0 or isinstance(id,str) and id.lower() in ("latest","recent","today"):
```
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
			elif id == 0 or isinstance(id, str) and id.lower() in ('latest', 'recent', 'today'): 
```
Single quotes and spaces for consistency.

File: cogs/fun.py (hunk 2 above)
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
			elif id is None:
				# random comic
				resp = await http.get('https://c.xkcd.com/random/comic')
			else:
				raise commands.CommandError('Invalid option: ' + str(id))
```
Unrecognized words don't make sense here imo.

File: cogs/fun.py [hunk 5]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
-			bs = BeautifulSoup(content, 'html.parser')
-			brs = bs.find('div', attrs=dict(id='middleContainer'))
-			img = brs.find('img')
-
-			if img is None:
-				await ctx.send(url)
-				return
-
+				id = str(resp.url).lstrip('https://xkcd.com/').rstrip("/")
+				url = f"https://xkcd.com/{id}/info.0.json"
```
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
				id = str(resp.url).lstrip('https://xkcd.com/').rstrip('/')
				url = 'https://xkcd.com/{}/info.0.json'.format(id)
```
Don't use f-strings and use single quotes.

File: cogs/fun.py [hunk 6]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
+			resp = await http.get(url)
+			if resp.status == 404:
+				raise commands.CommandError('Comic does not exist.')
+			if resp.status != 200:
+				raise commands.CommandError('Request failed.')
+			comic_json = await resp.json()
+			if id != comic_json['num']:
+				id = comic_json['num']
+			comic_url = f'https://xkcd.com/{id}'
```
Comment (SimplyMika) [inline cogs/fun.py:456]:
```suggestion
			comic_url = 'https://xkcd.com/' + str(id)
```
Don't use f-strings.

File: cogs/fun.py [hunk 7]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
-			e.set_image(url='https:' + img['src'])
-			e.set_footer(text=comic_url.lstrip('https://').rstrip('/'), icon_url='https://i.imgur.com/onzWnfd.png')
+				title=comic_json['safe_title'],
+				description=f"*{comic_json['alt']}*",
+				color=discord.Color.dark_theme(),
+				url=comic_url
+				)
+			comic_date = date(int(comic_json['year']),int(comic_json['month']),int(comic_json['day']))
+			e.set_image(url=comic_json['img'])
+			e.set_footer(text=f"{comic_url.lstrip('https://').rstrip('/')} • {str(comic_date)}", icon_url='https://i.imgur.com/onzWnfd.png')
```
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
				title=comic_json['safe_title'],
				description='*' + str(comic_json['alt']) + '*',
				color=discord.Color.dark_theme(),
				url=comic_url
			)
			comic_date = date(int(comic_json['year']), int(comic_json['month']), int(comic_json['day']))
			e.set_image(url=comic_json['img'])
			e.set_footer(text=comic_url.lstrip('https://').rstrip('/') + ' • ' + str(comic_date), icon_url='https://i.imgur.com/onzWnfd.png')
```
Consistencyyyyy

File: cogs/fun.py [hunk 8]
```diff
@@ -425,42 +425,48 @@ async def bill(self, ctx):
...
-			brs = bs.find('div', attrs=dict(id='middleContainer'))
-			img = brs.find('img')
-
-			if img is None:
-				await ctx.send(url)
-				return
-
+				id = str(resp.url).lstrip('https://xkcd.com/').rstrip("/")
+				url = f"https://xkcd.com/{id}/info.0.json"
+			resp = await http.get(url)
```
Comment (SimplyMika) [inline cogs/fun.py]:
```suggestion
			resp = await ctx.http.get(url)
```
Again, this raises an error.

**Review by SimplyMika — CHANGES_REQUESTED** [review]
Lots of consistency errors. (Might have still missed some.)
Lots of actual exceptions being thrown (test your code :P)
If you want, I can guide you through setting up a local Ace environment to play with.

Overall some nice ideas though! I like it :)

File: cogs/hl.py [hunk 9]
```diff
@@ -58,8 +69,10 @@ async def hl(self, ctx, *, code):
 
 		if len(code) > 2000:
 			raise commands.CommandError('Code contents too long to paste.')
-
-		message = await ctx.send(code)
+		try:
+			message = await ctx.send(content=code,reference=ctx.message.reference)
+		except HTTPException as e:
```
Comment (SimplyMika) [inline cogs/hl.py]:
```py
Traceback (most recent call last):
  ...
  File "/home/ace/test/AceBot/cogs/hl.py", line 74, in hl
    except HTTPException as e:
NameError: name 'HTTPException' is not defined
```

```suggestion
		except discord.HTTPException as e:
```
But really, don't send another message upon an HTTP Error. HTTP Error usually means something has gone wrong with the connection, not with any parameters.

File: cogs/hl.py [hunk 10]
```diff
@@ -58,8 +69,10 @@ async def hl(self, ctx, *, code):
 
 		if len(code) > 2000:
 			raise commands.CommandError('Code contents too long to paste.')
-
-		message = await ctx.send(code)
+		try:
+			message = await ctx.send(content=code,reference=ctx.message.reference)
```
Comment (SimplyMika) [inline cogs/hl.py:73]: parameter `reference` on `discord.ext.commands.Context.send` is only available since Discord.py 1.6, but Ace is using 1.5.1 (see `requirements.txt`), so this isn't possible yet.
You would need to bump up the version of Discord.py and test if it breaks anything.

File: cogs/hl.py [hunk 11]
```diff
@@ -30,9 +30,20 @@ class Highlighter(AceMixin, commands.Cog):
 
 	@commands.command(aliases=['h1'])
 	@commands.bot_has_permissions(manage_messages=True, add_reactions=True)
-	async def hl(self, ctx, *, code):
+	async def hl(self, ctx, *, code: str = None):
 		'''Highlight some code.'''
-
+		if ctx.message.reference is not None:
+			# get the replied to message contents if there is one
+			ref_msg = ctx.message.reference.resolved
```
Comment (SimplyMika) [inline cogs/hl.py:37]: `discord.MessageReference.resolved` was added in version 1.6, so it's not available.

File: cogs/hl.py [hunk 12]
```diff
@@ -30,9 +30,20 @@ class Highlighter(AceMixin, commands.Cog):
...
+			if ref_msg.author.bot:
+				raise commands.CommandError('Can not paste bot messages.')
+			code = ref_msg.content
+		else:
+			# include spaces/tabs at the beginning
+			code = ctx.message.content[len(ctx.prefix) + 3:]
+			if len(code) == 0:
+				raise commands.CommandError('No code to paste.')
+
```
Comment (SimplyMika) [inline cogs/hl.py:46]:
```suggestion
		if code is not None:
			# get the replied to message contents if there is one
			ref_msg = ctx.message.reference.resolved
			if ref_msg.author.bot:
				return
			code = ref_msg.content
```
You're never actually using the parameter `code`.

**Review by SimplyMika — CHANGES_REQUESTED** [review]
I won't review the other two commits, since I can't test them properly (my local instance obviously isn't running on the AHK server and I don't want to set up an entire fake AHK server to test this.

File: cogs/hl.py (hunk 9 above)
Comment (onerandomusername) [reply cogs/hl.py]: What this does is prevents an edge case if the message we try to reply to is deleted before we send the message.
Discord won't let the message go through, so we try to send it normally.

[docs](https://discordpy.readthedocs.io/en/latest/api.html?highlight=abc%20messageable%20send#discord.abc.Messageable.send)

File: cogs/ahk/ahk.py [hunk 13]
```diff
@@ -304,9 +304,11 @@ async def cloudahk_call(self, ctx, code, lang='ahk'):
 			' No output.\n' if stdout == '' else '\n```autoit\n{0}\n```'.format(stdout),
 			'Timed out' if time is None else '{0:.1f} seconds'.format(time),
 		)
-
-		await ctx.send(out)
-
+		try:
+			await ctx.send(content=out,reference=ctx.message.reference)
```
Comment (CloakerSmoker) [inline cogs/ahk/ahk.py:308]: This change doesn't make much sense in my opinion. I think it is confusing to have a response that mentions one user (the one who invoked the command) in the body, while also including a reply to a different user (if the invoker was replying to someone else).

For example, reading this is a bit of a mess, and hard to follow:

image

I think that replacing the mention with a reply to the invoker would be a better option.

Comment (onerandomusername) [reply cogs/ahk/ahk.py:308]: > I think that replacing the mention with a reply to the invoker would be a better option.

That's what should happen.
```suggestion
			await ctx.send(content=out, reference=ctx.message.to_reference())
```
That's what the line of code should be.

File: cogs/hl.py [hunk 14]
```diff
@@ -30,9 +30,20 @@ class Highlighter(AceMixin, commands.Cog):
...
-	async def hl(self, ctx, *, code):
+	async def hl(self, ctx, *, code: str = None):
 		'''Highlight some code.'''
-
+		if ctx.message.reference is not None:
+			# get the replied to message contents if there is one
+			ref_msg = ctx.message.reference.resolved
+			if ref_msg.author.bot:
+				raise commands.CommandError('Can not paste bot messages.')
+			code = ref_msg.content
```
Comment (CloakerSmoker) [inline cogs/hl.py:40]: Assigning `code` here does nothing, as the following line overwrites any value of it:

https://github.com/Run1e/AceBot/blob/14288c53ee7432f417161813b4b075f07111bfad/cogs/hl.py#L50

Resulting in a code box only containing a language being sent.

(Yes, I have seen the suggested changes by Micha, but that code has the same problem)

**Comment by Run1e** [comment]
Please be more specific about what your changes do, and please test your code before making a PR so others don't have to test it for you. Separate changes belong in separate PRs.

Regarding setting up a development instance, I think @Micha-ohne-el is creating a guide of sorts for this.

Other notes:
- f-strings are fine.
- the xkcd changes should be implemented as a group with separate subcommands instead of a union argument.
- auto-format your code before committing it.
